Adding java scenario test for reading file

hukoyu opened the pull request:
Adding java scenario test for reading file

MainApp class uses FileReadSample class to read file input.txt and verifies content of the file to pass the java scenario.

@letmaik please review

hukoyu committed:
Adding java scenario test for reading file

letmaik commented on tests/languages/java-file/app/MainApp.java:
```diff
@@ -0,0 +1,19 @@
+		System.out.println(data);
+		
+		if (data.compareTo("This is the content of file\n") == 0) {
+		    System.out.println("TEST_PASSED");
```
Let's exit with exit code 1 in case of failure, or just throw an exception which has the same effect. Identifiers should only be a last resort in tests.

hukoyu replied:
Done. Added exit()

letmaik commented on tests/languages/java-file/app/FileReadSample.java:
```diff
@@ -0,0 +1,19 @@
+public class FileReadSample {
+	public  String readFile(String filePath) {
+	StringBuilder contentBuilder = new StringBuilder();
+		try (Stream<String> stream = Files.lines( Paths.get(filePath), StandardCharsets.UTF_8)) 
```
Inconsistent indentation.

hukoyu replied:
try-catch removed per below feedback. Indentation is fixed.

letmaik commented on tests/languages/java-file/app/FileReadSample.java:
```diff
@@ -0,0 +1,19 @@
+public class FileReadSample {
+	public  String readFile(String filePath) {
+	StringBuilder contentBuilder = new StringBuilder();
+		try (Stream<String> stream = Files.lines( Paths.get(filePath), StandardCharsets.UTF_8)) 
```
I would remove the try-catch and let it crash if something goes wrong.

hukoyu replied:
Done

letmaik commented on tests/languages/java-file/Makefile:
```diff
@@ -0,0 +1,36 @@
+	javac -source 1.8 -target 1.8 $(APP_ROOT)/*.java
+
+$(DISK_IMAGE): $(APP_ROOT)/$(PROG).class
+	${SGXLKL_DISK_TOOL} create --size=${IMAGE_SIZE} --alpine="iputils iproute2 unzip libstdc++ openjdk8-jre nss" --copy=./${APP_ROOT}/ ${DISK_IMAGE}
```
Could you use a Dockerfile instead of `--alpine`? With a Dockerfile you can then compile inside Docker. I'd like to move over to Docker images for all new tests and gradually for existing ones as well.

hukoyu replied:
Implemented using Dockerfile and verified it is working properly.

letmaik commented on tests/languages/java-file/app/input.txt:
```diff
@@ -0,0 +1 @@
+This is the content of file
```
Either `This is the content of the file` or `This is the content of input.txt`.

hukoyu replied:
Fixed the grammar of data and made it two lines.

hukoyu committed:
Applying PR feedback

hukoyu committed:
Using docker image to compile java app and to create disk image for enclave

hukoyu commented:
@letmaik applied all of your feedback. Ready to be reveiewed.

letmaik commented on tests/languages/java-file/Makefile:
```diff
@@ -0,0 +1,35 @@
+.PHONY: run run-hw run-sw clean
+
+$(DISK_IMAGE): $(MAINAPP_JAVA) $(FILEREADER_JAVA)
+	docker build -t java-app:dev .
```
This works, but in tests we should use the Dockerfile directly in sgx-lkl-disk to avoid creating tagged/named Docker images on the CI or dev machines. You can use `--docker=Dockerfile` and remove the docker command.

hukoyu replied:
Done

letmaik commented on tests/languages/java-file/app/MainApp.java:
```diff
@@ -0,0 +1,22 @@
+		System.out.println("Content of the file:");
+		System.out.println(data);
+		
+		if (data.compareTo("This is the first line of the file\nThis is the second line of the file\n") == 0) {
```
I was thinking of this:
```
	        if (data.compareTo("This is the first line of the file\nThis is the second line of the file\n") != 0) {
		    System.exit(1);
		}
```

Would that not be enough?

hukoyu replied:
Sure will do the necessary change.

hukoyu replied:
Done

letmaik reviewed (changes requested):
Looks great, nearly done!

hukoyu committed:
Applying more PR feedback

hukoyu commented:
@letmaik I applied latest feedback. Ready to be merged.

Pengpeng-Microsoft commented on tests/languages/java-file/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM openjdk:8-alpine
```
Please change file structure as java / java_file.

hukoyu replied:
Done. Created java_hello_world and java_file under java folder. Will add new java-* folders under here

hukoyu committed:
Putting all java_* tests under java folder. Reorginizing helloworld java app to use docker image

hukoyu commented:
@Pengpeng-Microsoft I applied your feedback. Could you please review and approve if looks good.

letmaik commented:
I think it would be better to remove the `java_` prefix in the subfolders since the test names include the whole folder hierarchy. While you're at it I would rename "file" to "read-file" to make it absolutely clear what the test is doing.

davidchisnall commented:
What does this add to our test coverage?  We already have a test that reading a file works (the hello world test does this) and a test that a JVM can start.  The JVM is reading a file via the same libc interfaces as the other test, so it's not obvious to me that this adds anything to our tests.  What is the motivation for adding this test?

hukoyu commented:
> What does this add to our test coverage? We already have a test that reading a file works (the hello world test does this) and a test that a JVM can start. The JVM is reading a file via the same libc interfaces as the other test, so it's not obvious to me that this adds anything to our tests. What is the motivation for adding this test?

@davidchisnall I was adding more test scenarios with java language to have more complex scenarios than existing hello world java app. This one reads a file, I was planning to add for multithreading, networking and so on to have different more complext samples. You mean we don't need to add more java scenario tests? Hello world is good enough?

hukoyu committed:
Moving java_file folder to read_file folder

hukoyu committed:
Moved java_hello_world folder to hello_world folder

hukoyu commented:
> I think it would be better to remove the `java_` prefix in the subfolders since the test names include the whole folder hierarchy. While you're at it I would rename "file" to "read-file" to make it absolutely clear what the test is doing.

Thanks @letmaik . Before test result title was just including the most inner folder name. Thanks fixing to have whole folder structure. I didn't notice that. 
- moved tests/language/java/java_file to tests/language/java/read_file
- moved tests/language/java/java_hello_world to tests/language/java/hello_world

letmaik reviewed (approved):
Looks good. @davidchisnall I agree with you that the new test probably won't exercise more of the guts of SGX-LKL compared to the existing tests, but I'm still in favour of more real-life application tests just to create more confidence. We can consolidate tests later on again if we think we have too many that are not testing distinctive things.

davidchisnall commented:
I don't think it's a good idea to add tests without understanding what they are testing and why they are in the test suite.

Is this an integration test or a unit test?  We have an integration test for starting a JVM and unit tests for I/O system calls / libc library calls.  If it's intended as an integration test, then the same purpose would be better served by a more complex off-the-shelf Java application (or, ideally, something like the OpenJDK test suite).  

Increasing the things that we do in Java without understanding their impact on SGX-LKL coverage will bloat our test suite, slowing our test cycles, but not deliver any benefit in exchange for this cost.  Spawning threads, for example, is not something I would expect to have any benefit: a hello world Java program that uses Java's printf is already testing threading, the class loader, and GC.  OpenJDK implements an N:M threading model, so we won't necessarily see any extra SGX-LKL-visible concurrency from more threads.  

We want more (nightly) integration tests that check that end-to-end workloads are working.

We want more (pre-merge) unit tests and regression that check that specific features are working.

This does not seem to be either of them.

letmaik commented:
@davidchisnall I opened https://github.com/lsds/sgx-lkl/issues/247 for that.

hukoyu commented:
@davidchisnall @letmaik @prp I definitely want to work on adding scenario tests that will add value, I don't want to waste my time on useless scenarios per your explanation. Then my suggestion is please create issues for test scenarios for different languages that will add value to be implemented so that we can implement them and add to scenario tests.

jxyang commented:
@davidchisnall Since different language runtimes exercise musl/lkl in different ways, and the way to increase language runtime coverages is to add more programs written the particular language, I do see the value of this particular test case. 

Do we need to understand what a test case is testing before we write it? Absolutely. In this case, we are testing Java language features, with little regard to how the language runtime is implemented in sgx-lkl. Ultimately, that's what the users of sgx-lkl care most (if they plan to write confidential applications in Java).
